File: ios/tsc/src/ios/widget/SnackbarImpl.ts
// start - imports

	
import CommandAttr from '../../widget/CommandAttr';
import IWidget from '../../widget/IWidget';
import ILayoutParam from '../../widget/ILayoutParam';
import {plainToClass, Type, Exclude, Expose, Transform} from "class-transformer";
import 'babel-polyfill';
import {Gravity} from '../../widget/TypeConstants';
import {ITranform, TransformerFactory} from '../../widget/TransformerFactory';
import {Event} from '../../app/Event';
import {MotionEvent} from '../../app/MotionEvent';
import {DragEvent} from '../../app/DragEvent';
import {KeyEvent} from '../../app/KeyEvent';
import { ScopedObject } from '../../app/ScopedObject';












// end - imports
import {ViewImpl} from './ViewImpl';
export abstract class SnackbarImpl<T> extends ViewImpl<T>{
	//start - body
	static initialize() {
    }	
	@Type(() => CommandAttr)
	@Expose({ name: "duration" })
	duration!:CommandAttr<number>| undefined;
	@Type(() => CommandAttr)
	@Expose({ name: "text" })
	text!:CommandAttr<string>| undefined;
	@Type(() => CommandAttr)
	@Expose({ name: "show" })
	show_!:CommandAttr<boolean>| undefined;
	@Type(() => CommandAttr)
	@Expose({ name: "actionText" })
	actionText!:CommandAttr<string>| undefined;
	@Type(() => CommandAttr)
	@Expose({ name: "actionTextColor" })
	actionTextColor!:CommandAttr<string>| undefined;
	@Type(() => CommandAttr)
	@Expose({ name: "maxInlineActionWidth" })
	maxInlineActionWidth!:CommandAttr<string>| undefined;
	@Type(() => CommandAttr)
	@Expose({ name: "textColor" })
	textColor!:CommandAttr<string>| undefined;
	@Type(() => CommandAttr)
	@Expose({ name: "onClick" })
	onClick!:CommandAttr<string>| undefined;
	@Type(() => CommandAttr)
	@Expose({ name: "background" })
	background!:CommandAttr<string>| undefined;

	@Exclude()
	protected thisPointer: T;	
	protected abstract getThisPointer(): T;
	reset() : T {	
		super.reset();
		this.duration = undefined;
		this.text = undefined;
		this.show_ = undefined;
		this.actionText = undefined;
		this.actionTextColor = undefined;
		this.maxInlineActionWidth = undefined;
		this.textColor = undefined;
		this.onClick = undefined;
		this.background = undefined;
		return this.thisPointer;
	}
	constructor(id: string, path: string[], event:  string) {
		super(id, path, event);
		this.thisPointer = this.getThisPointer();
	}
	

	public setDuration(value : number) : T {
		this.resetIfRequired();
		if (this.duration == null || this.duration == undefined) {
			this.duration = new CommandAttr<number>();
		}
		
		this.duration.setSetter(true);
		this.duration.setValue(value);
		this.orderSet++;
		this.duration.setOrderSet(this.orderSet);
		return this.thisPointer;
	}
		

	public setText(value : string) : T {
		this.resetIfRequired();
		if (this.text == null || this.text == undefined) {
			this.text = new CommandAttr<string>();
		}
		
		this.text.setSetter(true);
		this.text.setValue(value);
		this.orderSet++;
		this.text.setOrderSet(this.orderSet);
		return this.thisPointer;
	}
		

	public show(value : boolean) : T {
		this.resetIfRequired();
		if (this.show_ == null || this.show_ == undefined) {
			this.show_ = new CommandAttr<boolean>();
		}
		
		this.show_.setSetter(true);
		this.show_.setValue(value);
		this.orderSet++;
		this.show_.setOrderSet(this.orderSet);
		return this.thisPointer;
	}
		

	public setActionText(value : string) : T {
		this.resetIfRequired();
		if (this.actionText == null || this.actionText == undefined) {
			this.actionText = new CommandAttr<string>();
		}
		
		this.actionText.setSetter(true);
		this.actionText.setValue(value);
		this.orderSet++;
		this.actionText.setOrderSet(this.orderSet);
		return this.thisPointer;
	}
		

	public setActionTextColor(value : string) : T {
		this.resetIfRequired();
		if (this.actionTextColor == null || this.actionTextColor == undefined) {
			this.actionTextColor = new CommandAttr<string>();
		}
		
		this.actionTextColor.setSetter(true);
		this.actionTextColor.setValue(value);
		this.orderSet++;
		this.actionTextColor.setOrderSet(this.orderSet);
		return this.thisPointer;
	}
		

	public setMaxInlineActionWidth(value : string) : T {
		this.resetIfRequired();
		if (this.maxInlineActionWidth == null || this.maxInlineActionWidth == undefined) {
			this.maxInlineActionWidth = new CommandAttr<string>();
		}
		
		this.maxInlineActionWidth.setSetter(true);
		this.maxInlineActionWidth.setValue(value);
		this.orderSet++;
		this.maxInlineActionWidth.setOrderSet(this.orderSet);
		return this.thisPointer;
	}
		

	public setTextColor(value : string) : T {
		this.resetIfRequired();
		if (this.textColor == null || this.textColor == undefined) {
			this.textColor = new CommandAttr<string>();
		}
		
		this.textColor.setSetter(true);
		this.textColor.setValue(value);
		this.orderSet++;
		this.textColor.setOrderSet(this.orderSet);
		return this.thisPointer;
	}
		

	public setOnClick(value : string) : T {
		this.resetIfRequired();
		if (this.onClick == null || this.onClick == undefined) {
			this.onClick = new CommandAttr<string>();
		}
		
		this.onClick.setSetter(true);
		this.onClick.setValue(value);
		this.orderSet++;
		this.onClick.setOrderSet(this.orderSet);
		return this.thisPointer;
	}
		

	public setBackground(value : string) : T {
		this.resetIfRequired();
		if (this.background == null || this.background == undefined) {
			this.background = new CommandAttr<string>();
		}
		
		this.background.setSetter(true);
		this.background.setValue(value);
		this.orderSet++;
		this.background.setOrderSet(this.orderSet);
		return this.thisPointer;
	}
		
	//end - body

}
	
//start - staticinit

export class Snackbar extends SnackbarImpl<Snackbar> implements IWidget{
    getThisPointer(): Snackbar {
        return this;
    }
    
   	public getClass() {
		return Snackbar;
	}
	
   	constructor(id: string, path: string[], event: string) {
		super(id, path, event);	
	}
}

SnackbarImpl.initialize();

//end - staticinit

File: swt/src/main/java/com/ashera/snackbar/SnackbarImpl.java
package com.ashera.snackbar;


//start - imports
import java.util.*;

import r.android.content.*;
import r.android.graphics.*;
import r.android.os.Build;
import r.android.view.*;

import com.ashera.core.IFragment;
import com.ashera.converter.*;
import r.android.annotation.SuppressLint;

import com.ashera.widget.bus.*;
import com.ashera.widget.*;

import org.eclipse.swt.widgets.*;
import org.eclipse.swt.graphics.*;
import org.eclipse.swt.*;

import static com.ashera.widget.IWidget.*;
//end - imports
import com.ashera.layout.ViewImpl;
@SuppressLint("NewApi")
public class SnackbarImpl extends BaseWidget {
	//start - body
	private View viewStub;
	private Object pane;
	public final static String LOCAL_NAME = "com.ashera.snackbar.Snackbar"; 
	public final static String GROUP_NAME = "com.ashera.snackbar.Snackbar";
	
	
	@Override
	public void loadAttributes(String attributeName) {

		WidgetFactory.registerAttribute(localName, new WidgetAttribute.Builder().withName("duration").withType("int"));
		WidgetFactory.registerAttribute(localName, new WidgetAttribute.Builder().withName("text").withType("resourcestring"));
		WidgetFactory.registerAttribute(localName, new WidgetAttribute.Builder().withName("show").withType("boolean"));
		WidgetFactory.registerAttribute(localName, new WidgetAttribute.Builder().withName("actionText").withType("resourcestring"));
		WidgetFactory.registerAttribute(localName, new WidgetAttribute.Builder().withName("actionTextColor").withType("colorstate"));
		WidgetFactory.registerAttribute(localName, new WidgetAttribute.Builder().withName("maxInlineActionWidth").withType("dimension"));
		WidgetFactory.registerAttribute(localName, new WidgetAttribute.Builder().withName("textColor").withType("colorstate"));
		WidgetFactory.registerAttribute(localName, new WidgetAttribute.Builder().withName("onClick").withType("string"));
		WidgetFactory.registerAttribute(localName, new WidgetAttribute.Builder().withName("background").withType("drawable").withUiFlag(UPDATE_UI_REQUEST_LAYOUT_N_INVALIDATE));
		WidgetFactory.registerAttribute(localName, new WidgetAttribute.Builder().withName("attributeUnderTest").withType("string"));

	}
	
	public SnackbarImpl() {
		super(LOCAL_NAME, LOCAL_NAME);
	}	

	@Override
	public IWidget newInstance() {
		return new SnackbarImpl();
	}
	
	@SuppressLint("NewApi")
	@Override
	public void create(IFragment fragment, Map<String, Object> params) {
		super.create(fragment, params);
		
        viewStub = new ViewExt();
        pane = new Composite(ViewImpl.getParent(this), SWT.NONE);
        ViewImpl.nativeMakeFrame(pane, 0, 0, 0, 0);
		nativeCreate(fragment, params);	
	}
	
	public class ViewExt extends View implements com.ashera.model.IViewStub{
		@Override
		public void remeasure() {
			if (getFragment() != null) {
				getFragment().remeasure();
			}
		}
        private Map<String, IWidget> templates;
    	@Override
    	public r.android.view.View inflateView(java.lang.String layout) {
    		if (templates == null) {
    			templates = new java.util.HashMap<String, IWidget>();
    		}
    		IWidget template = templates.get(layout);
    		if (template == null) {
    			template = (IWidget) quickConvert(layout, "template");
    			templates.put(layout, template);
    		}
    		IWidget widget = template.loadLazyWidgets(SnackbarImpl.this.getParent());
    		return (View) widget.asWidget();
    	}
	}

    

	@Override
	@SuppressLint("NewApi")
	public void setAttribute(WidgetAttribute key, String strValue, Object objValue, ILifeCycleDecorator decorator) {		
		Object nativeWidget = asNativeWidget();

		switch (key.getAttributeName()) {
			case "duration": {
				


		setDuration(objValue);



			}
			break;
			case "text": {
				


		setText(objValue);



			}
			break;
			case "show": {
				


		show(objValue);



			}
			break;
			case "actionText": {
				


		setActionText(objValue);



			}
			break;
			case "actionTextColor": {
				


		setActionTextColor(objValue);



			}
			break;
			case "maxInlineActionWidth": {
				


		setMaxInlineActionWidth(objValue);



			}
			break;
			case "textColor": {
				


		setTextColor(objValue);



			}
			break;
			case "onClick": {
				


		 setOnClick(objValue);



			}
			break;
			case "background": {
				


		setBackground(objValue);



			}
			break;
			case "attributeUnderTest": {
				


		 ViewImpl.setAttribute(this,  key, strValue, objValue, decorator);



			}
			break;
		default:
			break;
		}
	}
	@Override
	public Object asWidget() {
		return viewStub;
	}
	
	@Override
	@SuppressLint("NewApi")
	public Object getAttribute(WidgetAttribute key, ILifeCycleDecorator decorator) {
		Object nativeWidget = asNativeWidget();
		switch (key.getAttributeName()) {
		}
		
		return null;
	}
	@SuppressLint("NewApi")
private static class OnClickListener implements View.OnClickListener, com.ashera.widget.IListener{
private IWidget w; private View view; private String strValue; private String action;
public String getAction() {return action;}
public OnClickListener(IWidget w, String strValue)  {
this.w = w; this.strValue = strValue;
}
public OnClickListener(IWidget w, String strValue, String action)  {
this.w = w; this.strValue = strValue;this.action=action;
}
public void onClick (View v){
    
	if (action == null || action.equals("onClick")) {
		// populate the data from ui to pojo
		w.syncModelFromUiToPojo("onClick");
	    java.util.Map<String, Object> obj = getOnClickEventObj(v);
	    String commandName =  (String) obj.get(EventExpressionParser.KEY_COMMAND_NAME);
	    
	    // execute command based on command type
	    String commandType = (String)obj.get(EventExpressionParser.KEY_COMMAND_TYPE);
		switch (commandType) {
		case "+":
		    if (EventCommandFactory.hasCommand(commandName)) {
		    	 EventCommandFactory.getCommand(commandName).executeCommand(w, obj, v);
		    }

			break;
		default:
			break;
		}
		
		if (obj.containsKey("refreshUiFromModel")) {
			Object widgets = obj.remove("refreshUiFromModel");
			com.ashera.layout.ViewImpl.refreshUiFromModel(w, widgets, true);
		}
		if (w.getModelUiToPojoEventIds() != null) {
			com.ashera.layout.ViewImpl.refreshUiFromModel(w, w.getModelUiToPojoEventIds(), true);
		}
		if (strValue != null && !strValue.isEmpty() && !strValue.trim().startsWith("+")) {
		    com.ashera.core.IActivity activity = (com.ashera.core.IActivity)w.getFragment().getRootActivity();
		    if (activity != null) {
		    	activity.sendEventMessage(obj);
		    }
		}
	}
    return;
}//#####

public java.util.Map<String, Object> getOnClickEventObj(View v) {
	java.util.Map<String, Object> obj = com.ashera.widget.PluginInvoker.getJSONCompatMap();
    obj.put("action", "action");
    obj.put("eventType", "click");
    obj.put("fragmentId", w.getFragment().getFragmentId());
    obj.put("actionUrl", w.getFragment().getActionUrl());
    obj.put("namespace", w.getFragment().getNamespace());
    
    if (w.getComponentId() != null) {
    	obj.put("componentId", w.getComponentId());
    }
    
    PluginInvoker.putJSONSafeObjectIntoMap(obj, "id", w.getId());
     
    
    // parse event info into the map
    EventExpressionParser.parseEventExpression(strValue, obj);
    
    // update model data into map
    w.updateModelToEventMap(obj, "onClick", (String)obj.get(EventExpressionParser.KEY_EVENT_ARGS));
    return obj;
}
}

	
	

	private int duration = com.google.android.material.snackbar.Snackbar.LENGTH_INDEFINITE;
	private com.google.android.material.snackbar.Snackbar snackbar;
	private void show(Object objValue) {
		if ((boolean) objValue) {
			if (snackbar != null) {
				snackbar.dismiss();
				snackbar = null;
			}
			snackbar = com.google.android.material.snackbar.Snackbar.make(viewStub, text, duration);
			
			if (actionText != null) {
				if (onClickAction == null) {
					onClickAction = "";
				}
				snackbar.setAction(actionText, new OnClickListener(this, onClickAction));
			}
			
			if (actionTextColor != null) {
				snackbar.setActionTextColor(actionTextColor);
			}
			
			if (textColor != null) {
				snackbar.setTextColor(textColor);
			}

			if (maxInlineActionWidth != 0) {
				snackbar.setMaxInlineActionWidth(maxInlineActionWidth);
			}
			
			setBackgroundAttributes();
			
			snackbar.dismiss();
			snackbar.show();
		} else {
			if (snackbar != null) {
				snackbar.dismiss();
				snackbar = null;
			}
		}
	}

	private String text;
	private void setText(Object objValue) {
		this.text = (String) objValue;
	}

	private String onClickAction;
	private void setOnClick(Object objValue) {
		onClickAction = (String) objValue;
	}
	private String actionText;
	private void setActionText(Object objValue) {
		actionText = (String) objValue;
	}

	private r.android.content.res.ColorStateList textColor;
	private void setTextColor(Object objValue) {
		textColor = (r.android.content.res.ColorStateList) objValue;
		
	}
	private int maxInlineActionWidth;
	private void setMaxInlineActionWidth(Object objValue) {
		maxInlineActionWidth = (int) objValue;
		
	}
	private r.android.content.res.ColorStateList actionTextColor;
	private void setActionTextColor(Object objValue) {
		actionTextColor = (r.android.content.res.ColorStateList) objValue;
		
	}
	private void setDuration(Object objValue) {
		duration = (int) objValue;
		
	}

	


    @Override
    public void setVisible(boolean b) {
        ((View)asWidget()).setVisibility(b ? View.VISIBLE : View.GONE);
    }
    @Override
    public Object asNativeWidget() {
        return pane;
        
    }
    @Override
	public Class getViewClass() {
		return View.class;
	}
	//end - body

    //start - viewcode
	private void nativeCreate(IFragment fragment, Map<String, Object> params) {
		addDellocHandler();
	}
	
	private final static String DELLOC_EVENT = com.ashera.widget.bus.Event.StandardEvents.dealloc.toString();
	@com.google.j2objc.annotations.WeakOuter
	class DallocHandler extends com.ashera.widget.bus.EventBusHandler {

		public DallocHandler(String type) {
			super(type);
		}

		@Override
		protected void doPerform(Object payload) {
			if (snackbar != null) {
				snackbar.dismissImmediate();
				snackbar = null;
			}		
		}
		
	}
	private void addDellocHandler() {
		fragment.getEventBus().on(DELLOC_EVENT, new DallocHandler(DELLOC_EVENT));
	}

	private r.android.graphics.drawable.Drawable background;
	private void setBackground(Object objValue) {
		background = (r.android.graphics.drawable.Drawable) objValue;
	}


	private void setBackgroundAttributes() {
		if (background != null) {
			snackbar.setBackground(background);
		}
	}
	//end - viewcode
}
